File: WordSquares.java
import java.util.ArrayList;
import java.util.List;

// Time Complexity : O(n^n);
// Space Complexity : O(n*l*n*l);  Trie Space(n*l -> for each word , n*l > for the list of words at each node
// Did this code successfully run on Leetcode : Yes


public class WordSquares {
    TrieNode root;
    class TrieNode{
        TrieNode[] children;
        List<String> startsWith;

        public TrieNode(){
            this.children = new TrieNode[26];
            this.startsWith = new ArrayList<>();
        }
    }

    private void insert(String word){
        TrieNode curr = root;
        for(char c : word.toCharArray()){
            if(curr.children[c - 'a'] == null){
                curr.children[c - 'a'] = new TrieNode();
            }
            curr = curr.children[c - 'a'];
            curr.startsWith.add(word);
        }
    }

    private List<String> search(String prefix){
        TrieNode curr = root;
        for(char c : prefix.toCharArray()){
            if(curr.children[c - 'a'] == null){
                return new ArrayList<>();
            }
            curr = curr.children[c - 'a'];
        }
        return curr.startsWith;
    }

    List<List<String>> result;
    List<String> path;
    public List<List<String>> wordSquares(String[] words) {

        result = new ArrayList<>();
        path = new ArrayList<>();
        root = new TrieNode();

        for(String word: words){   // insert all the words in the trie TC: O(nl)  SC:n*l*n*l
            insert(word);
        }

        for(String word : words){   // Take one word and go for all other words
            path.add(word);
            helper(words);
            path.remove(path.size()-1);
        }
        return result;
    }

    private void helper(String[] words){

        if(path.size() == words[0].length()){
            result.add(new ArrayList<>(path));
            return;
        }
        StringBuilder currPrefix = new StringBuilder();    // form the prefix to select the next word
        int idx = path.size();
        for(String choosen : path){
            currPrefix.append(choosen.charAt(idx));
        }

        List<String> matchingWords = search(currPrefix.toString());
        for(String word: matchingWords){
            path.add(word);
            helper(words);
            path.remove(path.size()-1);
        }
    }
}

File: wordSquares.cpp
// Time complexity: 
// Space complexity: 

/*
Create a trie to store all the words and the startsWith vectors in each node of the trie 

Pick a new word and spawn a method to get the word square starting from that - use backtracking

Use a backrtacking method to get all the other words which satisy the condition of word squares and return the list of words when the length of the list reaches the lenght of the first word we were looking at 
*/


class Solution {
public:
    class TrieNode {
    public:
        vector<TrieNode*> children;
        vector<string> startsWith;
        TrieNode () {
            children = vector<TrieNode*> (26);
            startsWith = vector<string> ();
        }
    };
    TrieNode* root = new TrieNode();
    
    void constructTrie(string word) {
        TrieNode* curr = root;
        for(char c:word) {
            if(curr->children[c-'a'] == NULL) {
                curr->children[c-'a'] = new TrieNode();
            }
            curr = curr->children[c-'a'];
            curr->startsWith.push_back(word);
        }
        return;
    }
    
    vector<string> getWords(string prefix) {
        TrieNode* curr = root;
        for(char c: prefix) {
            if(curr->children[c-'a'] != NULL)
                curr = curr->children[c-'a'];
            else
                return vector<string>();
        }
        return curr->startsWith;
    }
    
    vector<vector<string>> result;
    

    // iterate over every word and use that as the starting point to get the qord squres
    vector<vector<string>> wordSquares(vector<string>& words) {
        if(words.size() == 0)
            return vector<vector<string>> ();
        // construct trie 
        for(string word: words)
            constructTrie(word);
        
        result = vector<vector<string>>();
        vector<string> list;
        
        // get result for each word and add it to the result
        for(string word: words) {
            //action
            list.push_back(word);
            //recurse
            backtrack(list, word.length());
            //backtrack
            list.pop_back();
        }
        return result;
    }
    
    // With the words currently in the list, get the prefix 
    // get the words starting with prefix and add each of those to the list and call the same method to continue getting the solution 
    // base case is when we have all the words we want in the list - 
    // The list can't have more words than the length of the frist word
    void backtrack(vector<string> list, int k) {
        // base
        // when the number of words in the list is k, add the list to the result 
        if(list.size() == k) {
            result.push_back(list);
            return;
        }
        // logic
        // pick up a string starting with prefix
        // get the prefix
        string prefix = "";
        int index = list.size();
        for(string word:list) {
            prefix += word[index];
        }
        //get the words starting with prefix (could be more than one)
        vector<string> startsWith = getWords(prefix);
        for(string word: startsWith) {
            // add each word to the list and call backtrack
            // action
            list.push_back(word);
            // recurse
            backtrack(list, word.length());
            // backtrack
            list.pop_back();
        }
        
    }
};